### PR TITLE
*: fix data races in from MockTiFlash

### LINKED ISSUE #31830: DATA RACE in the MockTiFlash.SyncStatus
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
==================
WARNING: DATA RACE
Read at 0x00c07501e060 by goroutine 1150:
  runtime.mapaccess2_fast64()
      /usr/local/go/src/runtime/map_fast64.go:52 +0x0
  github.com/pingcap/tidb/domain/infosync.(*MockTiFlash).setUpMockTiFlashHTTPServer.func1()
      /go/tidb/domain/infosync/tiflash_manager.go:268 +0xcd
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2046 +0x4d
  github.com/gorilla/mux.(*Router).ServeHTTP()
      /home/prow/go/pkg/mod/github.com/gorilla/mux@v1.8.0/mux.go:210 +0x366
  net/http.serverHandler.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2878 +0x89a
  net/http.(*conn).serve()
      /usr/local/go/src/net/http/server.go:1929 +0x12e4
  net/http.(*Server).Serve·dwrap·87()
      /usr/local/go/src/net/http/server.go:3033 +0x58
Previous write at 0x00c07501e060 by goroutine 256:
  runtime.mapassign_fast64()
      /usr/local/go/src/runtime/map_fast64.go:92 +0x0
  github.com/pingcap/tidb/domain/infosync.(*MockTiFlash).HandleSetPlacementRule.func1()
      /go/tidb/domain/infosync/tiflash_manager.go:322 +0x630
  github.com/pingcap/tidb/domain/infosync.(*MockTiFlash).HandleSetPlacementRule()
      /go/tidb/domain/infosync/tiflash_manager.go:335 +0x472
  github.com/pingcap/tidb/domain/infosync.(*mockTiFlashPlacementManager).SetPlacementRule()
      /go/tidb/domain/infosync/tiflash_manager.go:417 +0x217
  github.com/pingcap/tidb/domain/infosync.ConfigureTiFlashPDForTable()
      /go/tidb/domain/infosync/info.go:1016 +0x2a3
  github.com/pingcap/tidb/ddl.(*worker).onSetTableFlashReplica()
      /go/tidb/ddl/table.go:1104 +0x77a
  github.com/pingcap/tidb/ddl.(*worker).runDDLJob()
      /go/tidb/ddl/ddl_worker.go:835 +0xacd
  github.com/pingcap/tidb/ddl.(*worker).handleDDLJobQueue.func1()
      /go/tidb/ddl/ddl_worker.go:560 +0x4d1
  github.com/pingcap/tidb/kv.RunInNewTxn()
      /go/tidb/kv/txn.go:49 +0x142
  github.com/pingcap/tidb/ddl.(*worker).handleDDLJobQueue()
      /go/tidb/ddl/ddl_worker.go:516 +0x1db
  github.com/pingcap/tidb/ddl.(*worker).start()
      /go/tidb/ddl/ddl_worker.go:199 +0x6e6
  github.com/pingcap/tidb/ddl.(*ddl).Start·dwrap·13()
      /go/tidb/ddl/ddl.go:386 +0x47
Goroutine 1150 (running) created at:
  net/http.(*Server).Serve()
      /usr/local/go/src/net/http/server.go:3033 +0x847
  net/http/httptest.(*Server).goServe.func1()
      /usr/local/go/src/net/http/httptest/server.go:308 +0xc4
Goroutine 256 (running) created at:
  github.com/pingcap/tidb/ddl.(*ddl).Start()
      /go/tidb/ddl/ddl.go:386 +0xa53
  github.com/pingcap/tidb/domain.(*Domain).Init()
      /go/tidb/domain/domain.go:829 +0xec5
  github.com/pingcap/tidb/session.(*domainMap).Get.func1()
      /go/tidb/session/tidb.go:83 +0x694
  github.com/pingcap/tidb/util.RunWithRetry()
      /go/tidb/util/misc.go:65 +0x93
  github.com/pingcap/tidb/session.(*domainMap).Get()
      /go/tidb/session/tidb.go:71 +0x39a
  github.com/pingcap/tidb/session.createSessionWithOpt()
      /go/tidb/session/session.go:2768 +0x75
  github.com/pingcap/tidb/session.createSession()
      /go/tidb/session/session.go:2764 +0x25c
  github.com/pingcap/tidb/session.BootstrapSession()
      /go/tidb/session/session.go:2599 +0x245
  github.com/pingcap/tidb/ddl_test.(*tiflashDDLTestSuite).SetUpSuite()
      /go/tidb/ddl/ddl_tiflash_test.go:87 +0x2a5
  runtime.call16()
      /usr/local/go/src/runtime/asm_amd64.s:625 +0x48
  reflect.Value.Call()
      /usr/local/go/src/reflect/value.go:339 +0xd7
  github.com/pingcap/check.(*suiteRunner).runFixture.func1()
      /home/prow/go/pkg/mod/github.com/pingcap/check@v0.0.0-20200212061837-5e12011dc712/check.go:799 +0x245
  github.com/pingcap/check.(*suiteRunner).forkCall.func1()
      /home/prow/go/pkg/mod/github.com/pingcap/check@v0.0.0-20200212061837-5e12011dc712/check.go:739 +0x12b
==================
```

### PR DESCRIPTION
**Pull request by CalvinNeo**
### What problem does this PR solve?

Issue Number: close #31830 

Problem Summary:

Some fields in MockTiFlash is tested to have data race.

### What is changed and how it works?

Add lock for MockTiFlash.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
fix data races in from MockTiFlash
```

### CODE REVIEW

File: ddl/ddl_tiflash_test.go [hunk 1]
```diff
@@ -60,6 +60,8 @@ const (
 func (s *tiflashDDLTestSuite) SetUpSuite(c *C) {
```
Comment (hawkingrei) [inline ddl/ddl_tiflash_test.go:60]: You need migrate those new test case to new test framework.

Comment (hawkingrei) [reply ddl/ddl_tiflash_test.go:60]: ref #26022

**Review by xhebox — COMMENTED** [review]
I think it actually depends on #31832, wait for the merge of that first.

File: ddl/ddl_tiflash_test.go (hunk 1 above)
Comment (CalvinNeo) [reply ddl/ddl_tiflash_test.go:60]: https://github.com/pingcap/tidb/pull/31946 will do that

**Review by xhebox — CHANGES_REQUESTED** [review]
Resolve the conflicts first

**Review by xhebox — APPROVED** [review]

**Review by hawkingrei — APPROVED** [review]

**Pull request merged**